lib/Basic/Unix/TaskQueue.inc

```diff
@@ -555,34 +562,36 @@ static bool
 cleanUpAHungUpTask(Task &T,
                    const TaskQueue::TaskFinishedCallback FinishedCallback,
                    const TaskQueue::TaskSignalledCallback SignalledCallback) {
-  const Optional<int> StatusIfOK = waitForPid(T.getPid());
-  if (!StatusIfOK)
+  const auto StatusIfOK = waitForPid(T.getPid());
+  if (!StatusIfOK.first)
     return true;
 
   T.finishExecution();
-  int Status = *StatusIfOK;
+  int Status = *(StatusIfOK.first);
+  Optional<TaskFinishedMetrics> Metrics = StatusIfOK.second;
   return WIFEXITED(Status)
-             ? cleanUpAfterExit(Status, T, FinishedCallback)
+             ? cleanUpAfterExit(Status, T, Metrics, FinishedCallback)
              : WIFSIGNALED(Status)
-                   ? cleanUpAfterSignal(Status, T, SignalledCallback)
+                   ? cleanUpAfterSignal(Status, T, Metrics, SignalledCallback)
                    : false /* Can this case ever happen? */;
 }
 
-static Optional<int> waitForPid(const pid_t pidToWaitFor) {
+static std::pair<Optional<int>, Optional<TaskFinishedMetrics>> waitForPid(const pid_t pidToWaitFor) {
```

Comment:
Right now, you have waitForPid returning two optionals, but only two of the four possible cases can ever occur. If I'm right, it should be returning an optional pair, not a pair of optionals.

Reply:
It was an optional status before, that's why I chose to keep the status optional.

Reply:
Much better!